### PR TITLE
Use page statistics in Parquet reader

### PR DESCRIPTION
**Pull request by etseidl**
## Description
#14000 added the ability to write new page statistics to the Parquet writer. This PR uses these new statistics to avoid some string size computations. Benchmarks show an improvement in read times of up to 20%.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by etseidl** [comment]
parquet_read_decode benchmark
```
## [0] NVIDIA RTX A6000

|  data_type  |    io_type    |  cardinality  |  run_length  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |         Diff |   %Diff |
|-------------|---------------|---------------|--------------|------------|-------------|------------|-------------|--------------|---------|
|  INTEGRAL   | DEVICE_BUFFER |       0       |      1       |  14.942 ms |       0.74% |  14.989 ms |       0.76% |    46.778 us |   0.31% |
|  INTEGRAL   | DEVICE_BUFFER |     1000      |      1       |  14.115 ms |       0.50% |  14.201 ms |       0.46% |    85.853 us |   0.61% |
|  INTEGRAL   | DEVICE_BUFFER |       0       |      32      |  12.908 ms |       0.95% |  13.012 ms |       0.41% |   103.184 us |   0.80% |
|  INTEGRAL   | DEVICE_BUFFER |     1000      |      32      |  12.272 ms |       1.04% |  12.323 ms |       0.49% |    51.350 us |   0.42% |
|   STRING    | DEVICE_BUFFER |       0       |      1       |  46.516 ms |       0.45% |  45.672 ms |       0.50% |  -844.320 us |  -1.82% |
|   STRING    | DEVICE_BUFFER |     1000      |      1       |  13.762 ms |       0.68% |  13.015 ms |       0.71% |  -746.030 us |  -5.42% |
|   STRING    | DEVICE_BUFFER |       0       |      32      |  46.612 ms |       0.50% |  45.692 ms |       0.35% |  -920.004 us |  -1.97% |
|   STRING    | DEVICE_BUFFER |     1000      |      32      |  12.815 ms |       1.11% |  10.469 ms |       1.45% | -2345.521 us | -18.30% |
|    LIST     | DEVICE_BUFFER |       0       |      1       |  76.023 ms |       0.11% |  75.662 ms |       0.05% |  -361.155 us |  -0.48% |
|    LIST     | DEVICE_BUFFER |     1000      |      1       |  64.130 ms |       0.20% |  63.906 ms |       0.22% |  -224.179 us |  -0.35% |
|    LIST     | DEVICE_BUFFER |       0       |      32      |  55.348 ms |       0.13% |  55.170 ms |       0.09% |  -177.805 us |  -0.32% |
|    LIST     | DEVICE_BUFFER |     1000      |      32      |  56.574 ms |       0.21% |  56.434 ms |       0.15% |  -139.456 us |  -0.25% |
|   STRUCT    | DEVICE_BUFFER |       0       |      1       |  48.973 ms |       2.36% |  49.407 ms |       2.21% |   433.937 us |   0.89% |
|   STRUCT    | DEVICE_BUFFER |     1000      |      1       |  30.172 ms |       0.78% |  30.140 ms |       0.73% |   -32.744 us |  -0.11% |
|   STRUCT    | DEVICE_BUFFER |       0       |      32      |  49.314 ms |       2.83% |  50.072 ms |       2.74% |   758.959 us |   1.54% |
|   STRUCT    | DEVICE_BUFFER |     1000      |      32      |  28.299 ms |       0.37% |  27.150 ms |       0.37% | -1148.298 us |  -4.06% |
```

**Comment by etseidl** [comment]
parquet_read_chunks benchmark
```
## [0] NVIDIA RTX A6000

|   T    |    io_type    |  cardinality  |  run_length  |  byte_limit  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |          Diff |   %Diff |
|--------|---------------|---------------|--------------|--------------|------------|-------------|------------|-------------|---------------|---------|
| STRING | DEVICE_BUFFER |       0       |      1       |      0       |  45.542 ms |       0.13% |  45.354 ms |       0.52% |   -187.811 us |  -0.41% |
| STRING | DEVICE_BUFFER |     1000      |      1       |      0       |  13.455 ms |       0.43% |  13.059 ms |       0.47% |   -396.648 us |  -2.95% |
| STRING | DEVICE_BUFFER |       0       |      32      |      0       |  45.564 ms |       0.17% |  45.433 ms |       0.31% |   -131.213 us |  -0.29% |
| STRING | DEVICE_BUFFER |     1000      |      32      |      0       |  12.537 ms |       1.16% |  10.671 ms |       4.28% |  -1865.298 us | -14.88% |
| STRING | DEVICE_BUFFER |       0       |      1       |    500000    | 225.866 ms |       0.04% | 184.424 ms |       0.09% | -41442.361 us | -18.35% |
| STRING | DEVICE_BUFFER |     1000      |      1       |    500000    |  77.381 ms |       0.35% |  74.072 ms |       0.27% |  -3309.740 us |  -4.28% |
| STRING | DEVICE_BUFFER |       0       |      32      |    500000    | 225.814 ms |       0.05% | 183.758 ms |       0.18% | -42056.079 us | -18.62% |
| STRING | DEVICE_BUFFER |     1000      |      32      |    500000    |  61.544 ms |       0.51% |  52.393 ms |       0.20% |  -9151.281 us | -14.87% |
```
and the same but forcing PLAIN encoding
```
|   T    |    io_type    |  cardinality  |  run_length  |  byte_limit  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |          Diff |   %Diff |
|--------|---------------|---------------|--------------|--------------|------------|-------------|------------|-------------|---------------|---------|
| STRING | DEVICE_BUFFER |       0       |      1       |      0       |  45.449 ms |       0.32% |  45.168 ms |       0.56% |   -281.006 us |  -0.62% |
| STRING | DEVICE_BUFFER |     1000      |      1       |      0       |  46.893 ms |       0.69% |  46.812 ms |       0.79% |    -80.940 us |  -0.17% |
| STRING | DEVICE_BUFFER |       0       |      32      |      0       |  45.666 ms |       0.47% |  45.352 ms |       0.25% |   -314.615 us |  -0.69% |
| STRING | DEVICE_BUFFER |     1000      |      32      |      0       |  24.542 ms |       0.16% |  24.602 ms |       0.38% |     59.229 us |   0.24% |
| STRING | DEVICE_BUFFER |       0       |      1       |    500000    | 227.223 ms |       0.04% | 183.026 ms |       0.13% | -44196.857 us | -19.45% |
| STRING | DEVICE_BUFFER |     1000      |      1       |    500000    | 230.145 ms |       0.11% | 185.712 ms |       0.13% | -44432.935 us | -19.31% |
| STRING | DEVICE_BUFFER |       0       |      32      |    500000    | 227.210 ms |       0.07% | 183.010 ms |       0.07% | -44200.238 us | -19.45% |
| STRING | DEVICE_BUFFER |     1000      |      32      |    500000    | 208.225 ms |       0.13% | 164.224 ms |       0.11% | -44001.523 us | -21.13% |
```

File: cpp/src/io/parquet/reader_impl_preprocess.cu [hunk 1]
```diff
@@ -307,6 +307,90 @@ void generate_depth_remappings(std::map<int, std::pair<std::vector<int>, std::ve
...
+ * @brief Set fields on the pages that can be derived from page indexes.
+ *
+ * This replaces some preprocessing steps, such as page string size calculation.
+ */
+void fill_in_page_info(host_span<ColumnChunkDesc> chunks,
+                       device_span<PageInfo> pages,
+                       rmm::cuda_stream_view stream)
+{
+  auto const num_pages = pages.size();
+  std::vector<page_index_info> page_info(num_pages);
```
Comment (etseidl) [inline cpp/src/io/parquet/reader_impl_preprocess.cu:366]: Might want to go back to `hostdevice_vector` here.

Comment (vuule) [reply cpp/src/io/parquet/reader_impl_preprocess.cu:366]: It's up to you. Not all users will immediately benefit from the pinned pool so I don't want to push hd_vector as a recommendation just yet.

Comment (etseidl) [reply cpp/src/io/parquet/reader_impl_preprocess.cu:366]: Ok, maybe I'll leave a note or todo here then.

File: cpp/src/io/parquet/reader_impl_preprocess.cu [hunk 2]
```diff
@@ -307,6 +307,90 @@ void generate_depth_remappings(std::map<int, std::pair<std::vector<int>, std::ve
...
+  int32_t num_rows;
+  int32_t chunk_row;
+  int32_t num_nulls;
+  int32_t num_valids;
+  int32_t str_bytes;
+};
+
+// functor to copy page_index_info into the PageInfo struct
+struct copy_page_info {
+  device_span<page_index_info> page_info;
```
Comment (vuule) [inline cpp/src/io/parquet/reader_impl_preprocess.cu:337]:
```suggestion
  device_span<page_index_info const> page_info;
```

File: cpp/src/io/parquet/reader_impl_preprocess.cu [hunk 3]
```diff
@@ -307,6 +307,90 @@ void generate_depth_remappings(std::map<int, std::pair<std::vector<int>, std::ve
...
+  device_span<PageInfo> pages;
```
Comment (vuule) [inline cpp/src/io/parquet/reader_impl_preprocess.cu:338]: The naming is so tangled here. One member is named like the type of the other one. 
Maybe `index_infos` or `page_indexes` (given that `PageInfo` member is called `pages`)?

File: cpp/src/io/parquet/reader_impl_preprocess.cu (hunk 1 above)
Comment (vuule) [inline cpp/src/io/parquet/reader_impl_preprocess.cu:366]: oh no, the `page_index_info page_info` and `PageInfo page` conundrum applies here as well. Same as above, if pages are dropping their info part of the type name, maybe page_info should as well and be page_index :shrug:

**Review by vuule — COMMENTED** [review]
I already review the previous version of this feature, but going over it again since I didn't quite approve it back then (I almost did, IIRC).
Few nitpicks so far.

File: cpp/src/io/parquet/reader_impl_helpers.cpp [hunk 4]
```diff
@@ -346,6 +367,140 @@ size_type aggregate_reader_metadata::calc_num_row_groups() const
...
+        }
+      }
+
+      // If none of the ifs above triggered, then we have neither histogram (likely the writer
+      // doesn't produce them, the r:0 d:1 case should have been handled above). The column index
+      // doesn't give us value counts, so we'll have to rely on the page headers. If the histogram
+      // info is missing or insufficient, then just return without modifying the row_group_info.
+      if (not pg_info.num_nulls.has_value() or not pg_info.num_valid.has_value()) { return; }
+
+      // if using older page indexes that lack size info, don't use
```
Comment (vuule) [inline cpp/src/io/parquet/reader_impl_helpers.cpp:493]:
```suggestion
      // if using older page indexes that lack size info, don't use `pg_info`
```

**Review by vuule — APPROVED** [review]
Looks good, just one comment that reads a bit odd to me.

File: cpp/src/io/parquet/reader_impl_helpers.cpp (hunk 4 above)
Comment (etseidl) [reply cpp/src/io/parquet/reader_impl_helpers.cpp:493]: Cleaned this up and added a more detailed TODO...will push once CI finishes

Comment (vuule) [reply cpp/src/io/parquet/reader_impl_helpers.cpp:493]: looking at a few other PRs, this might be an "_if_ CI finishes" situation :grimacing:

**Comment by etseidl** [comment]
Looks like #15020 broke this (db 3 ets 0 :open_mouth: :sob: :rofl:). Regrouping...

File: cpp/src/io/parquet/reader_impl_helpers.cpp [hunk 5]
```diff
@@ -346,6 +367,143 @@ size_type aggregate_reader_metadata::calc_num_row_groups() const
...
 
+// Copies info from the column and offset indexes into the passed in row_group_info.
+void aggregate_reader_metadata::column_info_for_row_group(row_group_info& rg_info,
+                                                          size_type chunk_start_row) const
+{
+  auto const& fmd = per_file_metadata[rg_info.source_index];
+  auto const& rg  = fmd.row_groups[rg_info.index];
+
+  std::vector<column_chunk_info> chunks;
+  chunks.resize(rg.columns.size());
```
Comment (PointKernel) [inline cpp/src/io/parquet/reader_impl_helpers.cpp:378]:
```suggestion
  std::vector<column_chunk_info> chunks(rg.columns.size());
```

**Review by PointKernel — APPROVED** [review]

**Pull request merged**